File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public bool HasCycle(ListNode head) {
        if (head.next == null || head.next.next == null)
            return false;

        ListNode fast = head;
        ListNode slow = head;

        while (fast.next != null && fast.next.next != null)
        {
            fast = fast.next.next;
            slow = slow.next;

            if (fast == slow)
                return true;
        }

        return false;
    }
}

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        prev = None

        while head is not None:
            temp = head.next
            head.next = prev
            prev = head
            head = temp

        return prev
        # if head is None:
        #     return None

        # return self.recursiveRev(None, head)

        
    # def recursiveRev(prev: Optional[ListNode], curr: Optional[ListNode]) -> Optional[ListNode]:
    #     if curr is None:
    #         return prev

    #     nxt = curr.next

    #     curr.next = prev

    #     return recursiveRev(curr, nxt)